File: blog/scripts/analysis.py
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go
from sklearn import datasets
import cvxpy as cp
import pandas as pd
import numpy as np
import seaborn as sns
from mpl_toolkits.axes_grid1 import make_axes_locatable
from scipy.special import expit as inverse_logit
import itertools

def generate_data(n_samples=100, n_features=3, random_state=None, n_classes=2):
    if random_state is not None:
        np.random.seed(random_state)
    X, y = datasets.make_classification(n_samples=n_samples, n_features=n_features,
                                         n_informative=n_features, n_redundant=0, n_classes=n_classes,
                                         n_clusters_per_class=1, random_state=random_state, hypercube=True)
    return X, y.astype(int)

def plot_lda_3d_from_model(lda_model, X, y, feature_names=None, grid_size=30):
    """
    Plot 3D scatter and pairwise LDA decision planes using only lda_model.classes, .mus, .pis, .S.
    X: array (n_samples, 3)
    y: array (n_samples,)
    feature_names: list of length 3 for axis labels (optional)
    grid_size: number of points per axis in meshgrid
    """
    # Check 3D
    
    classes = lda_model.classes      # array of labels, length K
    mus = lda_model.mus              # shape (K, 3)
    pis = lda_model.pis              # shape (K,)
    S = lda_model.S                  # (3,3)
    # Invert S once
    invS = np.linalg.inv(S)
    
    # Prepare grid in feature1-feature2 plane
    f1 = np.linspace(X[:,0].min() - 1, X[:,0].max() + 1, grid_size)
    f2 = np.linspace(X[:,1].min() - 1, X[:,1].max() + 1, grid_size)
    xx, yy = np.meshgrid(f1, f2)
    
    fig = go.Figure()
    K = len(classes)
    
    # For each pair of classes, compute w and b for the plane
    for (i, j) in itertools.combinations(range(K), 2):
        mu_i = mus[i]
        mu_j = mus[j]
        pi_i = pis[i]
        pi_j = pis[j]
        # w_ij = invS @ (mu_i - mu_j)
        diff_mu = mu_i - mu_j
        w_ij = invS.dot(diff_mu)  # shape (3,)
        # compute quadratic terms mu^T invS mu
        qi = mu_i @ invS.dot(mu_i)
        qj = mu_j @ invS.dot(mu_j)
        b_ij = -0.5 * (qi - qj) + np.log(pi_i / pi_j)
        # Solve for z if w_z not too small
        if np.abs(w_ij[2]) < 1e-6:
            # skip nearly vertical plane in z
            continue
        zz = (-(w_ij[0] * xx + w_ij[1] * yy + b_ij) / w_ij[2])
        fig.add_trace(go.Surface(
            x=xx, y=yy, z=zz,
            opacity=0.5,
            showscale=False,
            name=f"Boundary {classes[i]} vs {classes[j]}",
            colorscale=[[0,'lightgray'], [1,'lightgray']],
            hoverinfo='skip'
        ))
    
    # Scatter for each class
    colors = px.colors.qualitative.Set1[:len(classes)]
    for idx, cls in enumerate(classes):
        mask = (y == cls)
        fig.add_trace(go.Scatter3d(
            x=X[mask, 0],
            y=X[mask, 1],
            z=X[mask, 2],
            mode='markers',
            name=f"Class {cls}",
            marker=dict(
                size=4,
                color=colors[idx],
            )
        ))
    
    # Axis labels
    labels = feature_names if feature_names is not None else ['Feature 1','Feature 2','Feature 3']
    # Compute symmetric ranges for equal scaling (optional but often helpful)
    x_min, x_max = X[:,0].min(), X[:,0].max()
    y_min, y_max = X[:,1].min(), X[:,1].max()
    z_min, z_max = X[:,2].min(), X[:,2].max()
    span = max(x_max - x_min, y_max - y_min, z_max - z_min) / 2
    x_mid = (x_max + x_min) / 2
    y_mid = (y_max + y_min) / 2
    z_mid = (z_max + z_min) / 2
    fig.update_layout(
        scene=dict(
            xaxis=dict(title=labels[0], range=[x_mid - span, x_mid + span]),
            yaxis=dict(title=labels[1], range=[y_mid - span, y_mid + span]),
            zaxis=dict(title=labels[2], range=[z_mid - span, z_mid + span]),
            aspectmode='cube'  # equal scaling on x,y,z
        ),
        title="3D LDA Decision Planes"
    )
    fig.show()
    return fig

def plot_qda_3d_with_isosurfaces(qda_model, X, y, feature_names=None, grid_size=30):
    """
    Plot 3D scatter and QDA decision boundaries as isosurfaces, using qda_model.discriminant_function.
    
    Parameters:
    -----------
    qda_model: fitted QDA instance (with .classes, .fit(), .discriminant_function(X), .predict(X))
    X: array-like of shape (n_samples, 3)
    y: array-like of shape (n_samples,)
    feature_names: list of length 3 for axis labels (optional)
    grid_size: int, number of grid points per axis
    """
    fig_list = []
    X = np.asarray(X)
    if X.shape[1] != 3:
        raise ValueError("X must have shape (n_samples, 3).")
    classes = np.asarray(qda_model.classes)
    K = len(classes)
    
    # Grid bounds with margin
    x_min, x_max = X[:,0].min() - 1, X[:,0].max() + 1
    y_min, y_max = X[:,1].min() - 1, X[:,1].max() + 1
    z_min, z_max = X[:,2].min() - 1, X[:,2].max() + 1
    
    # Build meshgrid
    xs = np.linspace(x_min, x_max, grid_size)
    ys = np.linspace(y_min, y_max, grid_size)
    zs = np.linspace(z_min, z_max, grid_size)
    XX, YY, ZZ = np.meshgrid(xs, ys, zs, indexing='ij')
    pts = np.vstack([XX.ravel(), YY.ravel(), ZZ.ravel()]).T  # shape (N,3)
    
    # Compute discriminant values on all grid points at once
    # shape (N, K)
    disc_vals = qda_model.discriminant_function(pts)
    
    fig = go.Figure()
    base_colors = px.colors.qualitative.Set1
    
    # For each pair of classes, plot the zero-level isosurface of δ_i - δ_j

    class_pairs = list(itertools.combinations(range(K), 2))
    for idx, (i, j) in enumerate(class_pairs):
        f_ij = disc_vals[:, i] - disc_vals[:, j]
        color = base_colors[idx % len(base_colors)]
        fig.add_trace(go.Isosurface(
            x=pts[:,0], y=pts[:,1], z=pts[:,2],
            value=f_ij,
            isomin=0.0, isomax=0.0,
            surface_count=1,
            showscale=False,
            caps=dict(x_show=False, y_show=False, z_show=False),
            colorscale=[[0, color], [1, color]],
            opacity=0.3,
            name=f"Boundary: {classes[i]} vs {classes[j]}"
        ))
        # Make a separate plot for each class pair, in the corresponding color:
        fig_class = go.Figure()
        fig_class.add_trace(go.Isosurface(
            x=pts[:,0], y=pts[:,1], z=pts[:,2],
            value=f_ij,
            isomin=0.0, isomax=0.0,
            surface_count=1,
            showscale=False,
            caps=dict(x_show=False, y_show=False, z_show=False),
            colorscale=[[0, color], [1, color]],
            opacity=0.3,
            name=f"Boundary: {classes[i]} vs {classes[j]}"
        ))
        # Plot original data points, only for selected classes
        scatter_colors = base_colors
        fig_class.add_trace(go.Scatter3d(
            x=X[y == classes[i], 0], y=X[y == classes[i], 1], z=X[y == classes[i], 2],
            mode='markers',
            marker=dict(
                size=4,
                color=scatter_colors[i % len(scatter_colors)],
                opacity=1
            ),
            name=f'Class {classes[i]}'
        ))
        fig_class.add_trace(go.Scatter3d(
            x=X[y == classes[j], 0], y=X[y == classes[j], 1], z=X[y == classes[j], 2],
            mode='markers',
            marker=dict(
                size=4,
                color=scatter_colors[j % len(scatter_colors)],
                opacity=1
            ),
            name=f'Class {classes[j]}'
        ))
        fig_class.update_layout(
            scene=dict(
                xaxis_title=feature_names[0] if feature_names else 'Feature 1',
                yaxis_title=feature_names[1] if feature_names else 'Feature 2',
                zaxis_title=feature_names[2] if feature_names else 'Feature 3',
                aspectmode='cube'
            ),
            title=f"QDA Boundary: {classes[i]} vs {classes[j]}",
            showlegend=True
        )
        fig_list.append(fig_class)
        fig_class.show()


    
    # Plot original data
    scatter_colors = base_colors
    for idx_cls, cls in enumerate(classes):
        mask = (y == cls)
        fig.add_trace(go.Scatter3d(
            x=X[mask, 0], y=X[mask, 1], z=X[mask, 2],
            mode='markers',
            marker=dict(
                size=4,
                color=scatter_colors[idx_cls % len(scatter_colors)],
                opacity=1
            ),
            name=f'Class {cls}'
        ))
    
    # Equal axis scaling
    labels = feature_names if feature_names is not None else ['Feature 1','Feature 2','Feature 3']
    span = max(x_max - x_min, y_max - y_min, z_max - z_min) / 2
    x_mid, y_mid, z_mid = (x_min + x_max)/2, (y_min + y_max)/2, (z_min + z_max)/2
    fig.update_layout(
        scene=dict(
            xaxis=dict(title=labels[0], range=[x_mid-span, x_mid+span]),
            yaxis=dict(title=labels[1], range=[y_mid-span, y_mid+span]),
            zaxis=dict(title=labels[2], range=[z_mid-span, z_mid+span]),
            aspectmode='cube'
        ),
        title="3D QDA Decision Boundaries for All Classes",
        showlegend=True,
        legend=dict(x=1.1, y=0.5)
    )
    fig_list.add(fig)
    fig.show()
    return fig_list

class LDA():
    def __init__(self):
        self.classes = None
        self.mus = None
        self.pis = None
        self.S = None
        self.mu = None

    def fit(self, X, y):
        n_samples, n_features = X.shape
        self.classes = np.unique(y)
        self.mus = np.zeros((len(self.classes), n_features))
        self.pis = np.zeros(len(self.classes))
    


        # Separate the data into two classes
        
        for i in range(self.classes.size):
            # Compute the means of each class
            mu = np.sum(X[y == self.classes[i]], axis=0) / np.sum(y == self.classes[i])
            
            self.mus[i] = mu.ravel()
            self.pis[i] = np.sum(y == self.classes[i]) / n_samples
     
        # Compute the overall mean
        mu = self.mus.T @ self.pis
        self.mu = mu.ravel()
   

        # Compute the sample covariance matrix
        S = (X - mu).T @ (X - mu) / n_samples
        self.S = S
        return self

    def _log_likelihood(self, X, mu_r, pi_r):
  
        # center
        diff = X - mu_r           

        # invert and det              

        # quadratic term for each row: sum_jk diff[i,j] * invS[j,k] * diff[i,k]
        #                                                     
        quad = np.einsum('ij,jk,ik->i', diff, np.linalg.inv(self.S), diff)

        const = 1 / (2 * np.pi * np.sqrt(np.linalg.det(self.S)))

        return np.log(pi_r) + const - 0.5 * quad

    
    def decision_function(self, X):
        likelihoods = np.zeros((X.shape[0], self.classes.size))
        for i in range(self.classes.size):
            ll = self._log_likelihood(X, self.mus[i], self.pis[i])
            likelihoods[:, i] = ll
        return likelihoods
        
    
    def predict(self, X):
        return np.argmax(self.decision_function(X), axis = 1)
    
class QDA:
    def __init__(self):
        self.classes = None
        self.mus = None
        self.pis = None
        self.Ss = None
        self.mu = None
    def fit(self, X, y):
        n_samples, n_features = X.shape
        self.classes = np.unique(y)
        self.mus = np.zeros((len(self.classes), n_features))
        self.pis = np.zeros(len(self.classes))
        self.Ss = np.zeros((len(self.classes), n_features, n_features))

        for i in range(self.classes.size):
            # Compute the means of each class
            mu = np.sum(X[y == self.classes[i]], axis=0) / np.sum(y == self.classes[i])
            self.mus[i] = mu.ravel()
            self.pis[i] = np.sum(y == self.classes[i]) / n_samples
            
            # Compute the sample covariance matrix for each class
            S = (X[y == self.classes[i]] - mu).T @ (X[y == self.classes[i]] - mu) / np.sum(y == self.classes[i])
            self.Ss[i] = S

        return self
    
    def discriminant_function(self, X):
        likelihoods = np.zeros((X.shape[0], self.classes.size))
        for i in range(self.classes.size):
            mu_r = self.mus[i]
            pi_r = self.pis[i]
            S_r = self.Ss[i]
            
            # Center the data
            diff = X - mu_r
            
            # Quadratic term
            quad = np.einsum('ij,jk,ik->i', diff, np.linalg.inv(S_r), diff)
            
            const = -0.5 * np.log(np.linalg.det(S_r))
            
            likelihoods[:, i] = np.log(pi_r) + const - 0.5 * quad
        
        return likelihoods
    def predict(self, X):
        return np.argmax(self.discriminant_function(X), axis=1)

File: blog/scripts/analysis.js
// Import math.js for matrix operations
import * as math from 'https://cdn.jsdelivr.net/npm/mathjs@11.8.0/+esm';
import * as d3 from 'https://cdn.jsdelivr.net/npm/d3@7.9.0/+esm';

export async function generateData(nSamples = 100, nFeatures = 3, randomState = null, nClasses = 2, variance = 1) {
    // Set random seed if provided
    if (randomState !== null) {
        // e.g. using the random_state from user input:
        Math.seedrandom(randomState);

    }
    
    // Generate centers for each class from unit cube edges
    const centers = [];
    for (let i = 0; i < nClasses; i++) {
        const randomDim = math.randomInt(nFeatures);
        let center = math.randomInt([nFeatures]);
        center[randomDim] = math.random();
        centers.push(center);
    }
    
    // Generate samples around centers
    const X = [];
    const y = [];
    const samplesPerClass = Math.floor(nSamples / nClasses);
    
    for (let i = 0; i < nClasses; i++) {
        for (let j = 0; j < samplesPerClass; j++) {
            // Add random noise to center
            const noise = math.random([nFeatures], -math.random(), math.random());
            console.log(noise);
            console.log(centers[i])
            const sample = math.add(centers[i], noise);
            X.push(sample);
            y.push(i);
        }
    }
    
    return [X, y];
}

export class LDA {
    constructor() {
        this.classes = null;
        this.mus = null;
        this.pis = null;
        this.S = null;
        this.mu = null;
    }
    
    fit(X, y) {
        const nSamples = X.length;
        const nFeatures = X[0].length;
        this.classes = [...new Set(y)];
        this.mus = Array(this.classes.length).fill().map(() => Array(nFeatures).fill(0));
        this.pis = Array(this.classes.length).fill(0);
        
        // Compute means and priors for each class
        for (let i = 0; i < this.classes.length; i++) {
            const classIndices = y.map((val, idx) => val === this.classes[i] ? idx : -1).filter(idx => idx !== -1);
            const classData = classIndices.map(idx => X[idx]);
            
            // Class mean
            this.mus[i] = math.mean(classData, 0);
            // Class prior
            this.pis[i] = classIndices.length / nSamples;
        }
        
        // Overall mean
        this.mu = math.multiply(math.transpose(this.mus), this.pis);
        
        // Compute covariance matrix
        const centered = math.subtract(X, this.mu);
        this.S = math.divide(math.multiply(math.transpose(centered), centered), nSamples);
        
        return this;
    }
    
    predict(X) {
        const scores = this.decisionFunction(X);
        return scores.map(row => math.indexOf(row, math.max(row)));
    }
    
    decisionFunction(X) {
        const invS = math.inv(this.S);
        return X.map(x => {
            return this.classes.map((_, i) => {
                const diff = math.subtract(x, this.mus[i]);
                const quad = math.multiply(math.multiply(diff, invS), math.transpose(diff));
                const logPrior = math.log(this.pis[i]);
                return logPrior - 0.5 * quad;
            });
        });
    }
}

export class QDA {
    constructor() {
        this.classes = null;
        this.mus = null;
        this.pis = null;
        this.Ss = null;
        this.mu = null;
    }
    
    fit(X, y) {
        const nSamples = X.length;
        const nFeatures = X[0].length;
        this.classes = [...new Set(y)];
        this.mus = Array(this.classes.length).fill().map(() => Array(nFeatures).fill(0));
        this.pis = Array(this.classes.length).fill(0);
        this.Ss = Array(this.classes.length).fill().map(() => 
            Array(nFeatures).fill().map(() => Array(nFeatures).fill(0))
        );
        
        for (let i = 0; i < this.classes.length; i++) {
            const classIndices = y.map((val, idx) => val === this.classes[i] ? idx : -1).filter(idx => idx !== -1);
            const classData = classIndices.map(idx => X[idx]);
            
            // Class mean
            this.mus[i] = math.mean(classData, 0);
            // Class prior
            this.pis[i] = classIndices.length / nSamples;
            this.mu = math.multiply(math.transpose(this.mus), this.pis);
            
            // Class covariance
            const centered = math.subtract(classData, this.mus[i]);
            this.Ss[i] = math.divide(math.multiply(math.transpose(centered), centered), classIndices.length);
        }
        
        return this;
    }
    
    discriminantFunction(X) {
        return X.map(x => {
            return this.classes.map((_, i) => {
                const diff = math.subtract(x, this.mus[i]);
                const invS = math.inv(this.Ss[i]);
                const quad = math.multiply(math.multiply(diff, invS), math.transpose(diff));
                const logDet = math.log(math.det(this.Ss[i]));
                const logPrior = math.log(this.pis[i]);
                return logPrior - 0.5 * (logDet + quad);
            });
        });
    }
    
    predict(X) {
        const scores = this.discriminantFunction(X);
        return scores.map(row => math.indexOf(row, math.max(row)));
    }
}

function meshgrid(x, y) {
    const xx = [];
    const yy = [];
    for (let i = 0; i < y.length; i++) {
        xx.push([...x]);
        yy.push(Array(x.length).fill(y[i]));
    }
    return [xx, yy];
}

function linspace(start, stop, n) {
    const arr = [];
    const step = (stop - start) / (n - 1);
    for (let i = 0; i < n; i++) {
        arr.push(start + step * i);
    }
    return arr;
}

export async function plotLDA3d(lda, X, y) {
    // Create Plotly 3D scatter plot with decision boundaries
    const classes = [...new Set(y)];
    const traces = [];

    // Add scatter points for each class
    classes.forEach((cls, idx) => {
        const classPoints = X.filter((_, i) => y[i] === cls);
        traces.push({
            type: 'scatter3d',
            x: classPoints.map(p => p[0]),
            y: classPoints.map(p => p[1]),
            z: classPoints.map(p => p[2]),
            mode: 'markers',
            name: `Class ${cls}`,
            marker: {
                size: 4,
                color: `hsl(${(360 * idx) / classes.length}, 70%, 50%)`
            }
        });
    });

    // Prepare grid in feature1-feature2 plane
    const gridSize = 50;
    const xMin = math.min(X.map(p => p[0])) - 1;
    const xMax = math.max(X.map(p => p[0])) + 1;
    const yMin = math.min(X.map(p => p[1])) - 1;
    const yMax = math.max(X.map(p => p[1])) + 1;
    
    const f1 = linspace(xMin, xMax, gridSize);
    const f2 = linspace(yMin, yMax, gridSize);
    
    // Create meshgrid
    const [xx, yy] = meshgrid(f1, f2);
    
    // Get the inverse of the covariance matrix
    const invS = math.inv(lda.S);
    
    // For each pair of classes, compute w and b for the plane
    for (let i = 0; i < classes.length; i++) {
        for (let j = i + 1; j < classes.length; j++) {
            const mu_i = lda.mus[i];
            const mu_j = lda.mus[j];
            const pi_i = lda.pis[i];
            const pi_j = lda.pis[j];
            
            // Calculate w_ij = invS @ (mu_i - mu_j)
            const diff_mu = math.subtract(mu_i, mu_j);
            const w_ij = math.multiply(invS, diff_mu);
            
            // Compute quadratic terms mu^T invS mu
            const qi = math.multiply(math.multiply(mu_i, invS), math.transpose(mu_i));
            const qj = math.multiply(math.multiply(mu_j, invS), math.transpose(mu_j));
            const b_ij = -0.5 * (qi - qj) + math.log(pi_i / pi_j);
            
            // Skip if the plane is nearly vertical in z
            if (Math.abs(w_ij[2]) < 1e-6) continue;
            
            // Calculate z coordinates for the decision boundary surface
            const zz = xx.map((row, r) => 
                row.map((x, c) => 
                    (-(w_ij[0] * x + w_ij[1] * yy[r][c] + b_ij) / w_ij[2])
                )
            );
            
            // Add surface trace
            traces.push({
                type: 'surface',
                x: xx,
                y: yy,
                z: zz,
                opacity: 0.3,
                name: `Boundary ${classes[i]} vs ${classes[j]}`,
                colorscale: [[0, 'lightgray'], [1, 'lightgray']],
                showscale: false,
                hoverinfo: 'name',
                customdata: [{classes: [i, j]}],
                hoverlabel: {
                    bgcolor: 'white',
                    font: {color: 'black'}
                }
            });
        }
    }
    let x_min = Math.min(...X.map((r) => r[0]));
    let x_max = Math.max(...X.map((r) => r[0]));
    let y_min = Math.min(...X.map((r) => r[1]));
    let y_max = Math.max(...X.map((r) => r[1]));
    let z_min = Math.min(...X.map((r) => r[2]));
    let z_max = Math.max(...X.map((r) => r[2]));
    

    const span =  Math.max(x_max - x_min, y_max - y_min, z_max - z_min) / 2

    const x_mid = (x_max + x_min) / 2
    const y_mid = (y_max + y_min) / 2
    const z_mid = (z_max + z_min) / 2

    const layout = {
        title: '3D LDA Decision Boundaries',
        scene: {
            xaxis: {title: classes[0], range: [x_mid - span, x_mid + span]},
            yaxis: {title: classes[1], range: [y_mid - span, y_mid + span]},
            zaxis: {title: classes[2], range: [z_mid - span, z_mid + span]},
            aspectmode: 'cube'
        },
        margin: {l: 0, r: 0, b: 0, t: 30},
    };
    
    
    return {traces, layout};
}


export async function plotQDA3d(qda, X, y) {
    // Similar to LDA plot but with quadratic decision boundaries
    const classes = [...new Set(y)];
    const traces = [];
    const subtraces = [];
    const sub = [];
    // Add scatter points for each class
    classes.forEach((cls, idx) => {
        const classPoints = X.filter((_, i) => y[i] === cls);
        const trc = {
            type: 'scatter3d',
            x: classPoints.map(p => p[0]),
            y: classPoints.map(p => p[1]),
            z: classPoints.map(p => p[2]),
            mode: 'markers',
            name: `Class ${cls}`,
            marker: {
                size: 4,
                color: `hsl(${(360 * idx) / classes.length}, 70%, 50%)`
            }
        }
        traces.push(
            trc
        );
        sub.push(
            trc
        );
    });
    subtraces.push(sub);

    // Prepare grid in feature1-feature2 plane
    const gridSize = 50; // Increased grid density
    const xMin = math.min(X.map(p => p[0]));
    const xMax = math.max(X.map(p => p[0]));
    const yMin = math.min(X.map(p => p[1]));
    const yMax = math.max(X.map(p => p[1]));
    
    const f1 = linspace(xMin, xMax, gridSize);
    const f2 = linspace(yMin, yMax, gridSize);
    
    // Create meshgrid
    const [xx, yy] = meshgrid(f1, f2);

    const color = d3.scaleOrdinal(d3.schemeCategory10).domain(classes);

    // For each pair of classes, compute the quadratic decision boundary
    for (let i = 0; i < classes.length; i++) {
        for (let j = i + 1; j < classes.length; j++) {
            // First pass: compute  z values
            const zRange = linspace(X[0].length >= 3 ? math.min(X.map((v) => v[2])) : -1, X[0].length >= 3 ? math.max(X.map((v) => v[2])) : 1, gridSize); 
            const zz= xx.map((row, r) => 
                row.map((x, c) => {
                    
                    let minDiff = Infinity;
                    let bestZ = 0;
                  
                    zRange.forEach(z => {
                        let p = [x, yy[r][c], z];
                       
                        p = X[0].length > 3 ? p.concat(qda.mu.slice(3)) : p;
                        p = [p]
                
                        let discriminant = qda.discriminantFunction(p).flat();
                        const diff = math.abs(discriminant[i] - discriminant[j]);
                        
                        if (diff < minDiff) {
                            minDiff = diff;
                            bestZ = z;
                        }
                    });
                    
                    return minDiff < 0.5 && (bestZ != 0 || bestZ != undefined) ? bestZ : NaN;
                })
            );
            
        
            // Add surface trace
            const surface = {
                type: 'surface',
                x: xx,
                y: yy,
                z: zz,
                opacity: 0.3,
                showscale: false,
                name: `Boundary ${classes[i]} vs ${classes[j]}`,
                colorscale: [[0, color(i)], [1, color(j)]],
                hoverinfo: 'name',
                customdata: [{classes: [i, j]}],
                hoverlabel: {
                    bgcolor: 'white',
                    font: {color: 'black'}
                }
            }

            traces.push(surface);
            subtraces.push(surface);

            
        }
    }
    
    const layout = {
        title: '3D QDA Decision Boundaries',
        scene: {
            xaxis: {
            range: [-1, 1],
            },
            yaxis: {
            range: [-1, 1], 
            },
            zaxis: {
            range: [-1, 1],
            }
        },
        margin: {l:0,r:0,b:0,t:30}

    };
    
    return {traces, subtraces, layout};
}
